tools/snapshot/snapshot.go

```diff
@@ -0,0 +1,251 @@
+
+const (
+	CMD_SNAPSHOT Command = iota
+	CMD_RESTORE
+)
+
+func main() {
+	// Ensure there's a command specified.
+	// TODO - as we add more commands, we should probably use a library like spf13/cobra.
+	if len(os.Args) < 2 {
+		fmt.Println("Please specify whether to (b)ackup or (r)estore.")
+		os.Exit(1)
+	}
+
+	// Determine the command.
+	var command Command
+	switch os.Args[1] {
+	case "s", "snap", "snapshot":
+		command = CMD_SNAPSHOT
+	case "r", "restore":
+		command = CMD_RESTORE
+	default:
+		fmt.Println("Please specify whether to (s)snapshot or (r)estore.")
+	}
+
+	// Determine the path to the top-level directory (where the Makefile resides).
+	repoRoot, err := getRepoRoot()
+	if err != nil {
+		fmt.Printf("Error determining repo root: %v\n", err)
+		os.Exit(1)
+	}
+
+	// Change the working directory to the repo root.
+	if err := os.Chdir(repoRoot); err != nil {
+		fmt.Printf("Error changing directory to repo root: %v\n", err)
+		os.Exit(1)
+	}
+
+	// Get the home directory so we can get the snapshots dir.
+	homedir, err := os.UserHomeDir()
+	if err != nil {
+		fmt.Printf("Could not determine home directory: %v\n", err)
+		return
+	}
+
+	// Run the command.
+	switch command {
+	case CMD_SNAPSHOT:
+		snapshot(homedir)
+	case CMD_RESTORE:
+		restore(homedir)
+	}
+}
+
+// Restore a snapshot.
+func restore(homedir string) error {
+	snapshotsDir := filepath.Join(homedir, ".fleet", "snapshots")
+	_, err := os.Lstat(snapshotsDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			fmt.Printf("You don't currently have any snapshots.\n")
+		} else {
+			// Handle other PathError-specific cases
+			fmt.Printf("Error reading snapshots directory (%s): %v\n", snapshotsDir, err)
+		}
+		return err
+	}
+
+	// Walk the ~/.fleet/snapshots directory if it exists.
+	dirEntries, err := os.ReadDir(snapshotsDir)
+	var snapshots []Snapshot
+	for _, entry := range dirEntries {
+		if entry.IsDir() {
+			// Ensure there's a db backup file.
+			dbBackupFile := filepath.Join(snapshotsDir, entry.Name(), "db.sql.gz")
+			dbBackupFileInfo, err := os.Lstat(dbBackupFile)
+			if err != nil {
+				continue
+			}
+			snapshot := Snapshot{
+				Name: entry.Name(),
+				Date: dbBackupFileInfo.ModTime().Format("Jan 02, 2006 03:04:05 PM"),
+				Path: dbBackupFile,
+			}
+			snapshots = append(snapshots, snapshot)
+		}
+	}
+
+	// Set up and run the "Select snapshot" UI.
+	templates := &promptui.SelectTemplates{
+		Label:    "{{ .Name }}",
+		Active:   "> {{ .Name }} ({{ .Date }})",
+		Inactive: "{{ .Name }} ({{ .Date }})",
+		Selected: "{{ .Name }} ({{ .Date }})",
+	}
+	prompt := promptui.Select{
+		Label:     "Select snapshot to restore",
+		Items:     snapshots,
+		Templates: templates,
+	}
+	index, _, err := prompt.Run()
+	if err != nil {
+		fmt.Printf("Prompt failed %v\n", err)
+		return err
+	}
+
+	// Prepare the restore script with the selected snapshot.
+	cmd := exec.Command("/Users/scott/Development/fleet/tools/backup_db/restore.sh", snapshots[index].Path)
```

Comment:
👀

Reply:
Are we all getting access to your home directory?

Reply:
i mean if it works... 😆